# MACHINE LEARNING: REGRESSION - PREDICTING ENERGY EFFICIENCY OF BUILDINGS

## CHUKWUMA ANTHONY NWACHUKWU

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df=pd.read_csv('C:/Users/TonyC/Desktop/Intern Projects/Predicting energy efficiency of buildings/energydata_complete.csv')

In [3]:
df.head()

date  Appliances  lights     T1       RH_1    T2       RH_2  \
0  1/11/2016 17:00          60      30  19.89  47.596667  19.2  44.790000   
1  1/11/2016 17:10          60      30  19.89  46.693333  19.2  44.722500   
2  1/11/2016 17:20          50      30  19.89  46.300000  19.2  44.626667   
3  1/11/2016 17:30          50      40  19.89  46.066667  19.2  44.590000   
4  1/11/2016 17:40          60      40  19.89  46.333333  19.2  44.530000   

      T3       RH_3         T4  ...         T9   RH_9  T_out  Press_mm_hg  \
0  19.79  44.730000  19.000000  ...  17.033333  45.53   6.60        733.5   
1  19.79  44.790000  19.000000  ...  17.066667  45.56   6.48        733.6   
2  19.79  44.933333  18.926667  ...  17.000000  45.50   6.37        733.7   
3  19.79  45.000000  18.890000  ...  17.000000  45.40   6.25        733.8   
4  19.79  45.000000  18.890000  ...  17.000000  45.40   6.13        733.9   

   RH_out  Windspeed  Visibility  Tdewpoint        rv1        rv2  
0    92.0   7.000000   63.000000        5.3  13.275433  13.275433  
1    92.0   6.666667   59.166667        5.2  18.606195  18.606195  
2    92.0   6.333333   55.333333        5.1  28.642668  28.642668  
3    92.0   6.000000   51.500000        5.0  45.410390  45.410390  
4    92.0   5.666667   47.666667        4.9  10.084097  10.084097  

[5 rows x 29 columns]

In [4]:
#renames columns
column_names={'T1':'Temperature_in_Kitchen', 'T2': 'Temperature_in_Livingroom', 
             'T3':  'Temperature_in_Laundryroom', 'T4': 'Temperature_in_office',
              'T5': 'Temperature_in_Bathroom', 'T6': 'Temperature_outside_building',
              'T7':'Temperature_in_ironingroom', 'T8': 'Temperature_in_teenager_room',
              'T9': 'Temperature_in_Parent_room', 'RH_1':'Humidity_in_Kitchen',
              'RH_2': 'Humidity_in_Living_room', 'RH_3': 'Humidity_in_Laundary_room',
              'RH_4': 'Humidity_in_office', 'RH_5': 'Humidity_in_Bathroom',
              'RH_6': 'Humidity_outside_building', 'RH_7': 'Humidity_in_ironingroom', 
             'RH_8': 'Humidity_in_teenager_room','RH_9': 'Humidity_in_parent_room',
            'rv1': 'Random_variable_1','rv2': 'Random_variable_2', 'T_out': 'Temperature_Outside',
             'RH_out': 'Humidiity_Outside', 'Press_mm_hg': 'Pressure_mm_Hg'
             }
#command to assign new values to columns
df=df.rename(columns=column_names)

In [5]:
df.head()

date  Appliances  lights  Temperature_in_Kitchen  \
0  1/11/2016 17:00          60      30                   19.89   
1  1/11/2016 17:10          60      30                   19.89   
2  1/11/2016 17:20          50      30                   19.89   
3  1/11/2016 17:30          50      40                   19.89   
4  1/11/2016 17:40          60      40                   19.89   

   Humidity_in_Kitchen  Temperature_in_Livingroom  Humidity_in_Living_room  \
0            47.596667                       19.2                44.790000   
1            46.693333                       19.2                44.722500   
2            46.300000                       19.2                44.626667   
3            46.066667                       19.2                44.590000   
4            46.333333                       19.2                44.530000   

   Temperature_in_Laundryroom  Humidity_in_Laundary_room  \
0                       19.79                  44.730000   
1                       19.79                  44.790000   
2                       19.79                  44.933333   
3                       19.79                  45.000000   
4                       19.79                  45.000000   

   Temperature_in_office  ...  Temperature_in_Parent_room  \
0              19.000000  ...                   17.033333   
1              19.000000  ...                   17.066667   
2              18.926667  ...                   17.000000   
3              18.890000  ...                   17.000000   
4              18.890000  ...                   17.000000   

   Humidity_in_parent_room  Temperature_Outside  Pressure_mm_Hg  \
0                    45.53                 6.60           733.5   
1                    45.56                 6.48           733.6   
2                    45.50                 6.37           733.7   
3                    45.40                 6.25           733.8   
4                    45.40                 6.13           733.9   

   Humidiity_Outside  Windspeed  Visibility  Tdewpoint  Random_variable_1  \
0               92.0   7.000000   63.000000        5.3          13.275433   
1               92.0   6.666667   59.166667        5.2          18.606195   
2               92.0   6.333333   55.333333        5.1          28.642668   
3               92.0   6.000000   51.500000        5.0          45.410390   
4               92.0   5.666667   47.666667        4.9          10.084097   

   Random_variable_2  
0          13.275433  
1          18.606195  
2          28.642668  
3          45.410390  
4          10.084097  

[5 rows x 29 columns]

In [6]:
df.isnull().sum()

date                            0
Appliances                      0
lights                          0
Temperature_in_Kitchen          0
Humidity_in_Kitchen             0
Temperature_in_Livingroom       0
Humidity_in_Living_room         0
Temperature_in_Laundryroom      0
Humidity_in_Laundary_room       0
Temperature_in_office           0
Humidity_in_office              0
Temperature_in_Bathroom         0
Humidity_in_Bathroom            0
Temperature_outside_building    0
Humidity_outside_building       0
Temperature_in_ironingroom      0
Humidity_in_ironingroom         0
Temperature_in_teenager_room    0
Humidity_in_teenager_room       0
Temperature_in_Parent_room      0
Humidity_in_parent_room         0
Temperature_Outside             0
Pressure_mm_Hg                  0
Humidiity_Outside               0
Windspeed                       0
Visibility                      0
Tdewpoint                       0
Random_variable_1               0
Random_variable_2               0
dtype: int64

In [7]:
df.describe()

Appliances        lights  Temperature_in_Kitchen  \
count  19735.000000  19735.000000            19735.000000   
mean      97.694958      3.801875               21.686571   
std      102.524891      7.935988                1.606066   
min       10.000000      0.000000               16.790000   
25%       50.000000      0.000000               20.760000   
50%       60.000000      0.000000               21.600000   
75%      100.000000      0.000000               22.600000   
max     1080.000000     70.000000               26.260000   

       Humidity_in_Kitchen  Temperature_in_Livingroom  \
count         19735.000000               19735.000000   
mean             40.259739                  20.341219   
std               3.979299                   2.192974   
min              27.023333                  16.100000   
25%              37.333333                  18.790000   
50%              39.656667                  20.000000   
75%              43.066667                  21.500000   
max              63.360000                  29.856667   

       Humidity_in_Living_room  Temperature_in_Laundryroom  \
count             19735.000000                19735.000000   
mean                 40.420420                   22.267611   
std                   4.069813                    2.006111   
min                  20.463333                   17.200000   
25%                  37.900000                   20.790000   
50%                  40.500000                   22.100000   
75%                  43.260000                   23.290000   
max                  56.026667                   29.236000   

       Humidity_in_Laundary_room  Temperature_in_office  Humidity_in_office  \
count               19735.000000           19735.000000        19735.000000   
mean                   39.242500              20.855335           39.026904   
std                     3.254576               2.042884            4.341321   
min                    28.766667              15.100000           27.660000   
25%                    36.900000              19.530000           35.530000   
50%                    38.530000              20.666667           38.400000   
75%                    41.760000              22.100000           42.156667   
max                    50.163333              26.200000           51.090000   

       ...  Temperature_in_Parent_room  Humidity_in_parent_room  \
count  ...                19735.000000             19735.000000   
mean   ...                   19.485828                41.552401   
std    ...                    2.014712                 4.151497   
min    ...                   14.890000                29.166667   
25%    ...                   18.000000                38.500000   
50%    ...                   19.390000                40.900000   
75%    ...                   20.600000                44.338095   
max    ...                   24.500000                53.326667   

       Temperature_Outside  Pressure_mm_Hg  Humidiity_Outside     Windspeed  \
count         19735.000000    19735.000000       19735.000000  19735.000000   
mean              7.412580      755.522602          79.750418      4.039752   
std               5.318464        7.399441          14.901088      2.451221   
min              -5.000000      729.300000          24.000000      0.000000   
25%               3.670000      750.933333          70.333333      2.000000   
50%               6.920000      756.100000          83.666667      3.666667   
75%              10.400000      760.933333          91.666667      5.500000   
max              26.100000      772.300000         100.000000     14.000000   

         Visibility     Tdewpoint  Random_variable_1  Random_variable_2  
count  19735.000000  19735.000000       19735.000000       19735.000000  
mean      38.330834      3.760995          24.988033          24.988033  
std       11.794719      4.195248          14.496634          14.496634  
min        1.000000     -6.600000           0.005322          

In [8]:
#removing dropping features date and lights
new=df.drop(columns=['lights','date'])

In [9]:
new.head()

Appliances  Temperature_in_Kitchen  Humidity_in_Kitchen  \
0          60                   19.89            47.596667   
1          60                   19.89            46.693333   
2          50                   19.89            46.300000   
3          50                   19.89            46.066667   
4          60                   19.89            46.333333   

   Temperature_in_Livingroom  Humidity_in_Living_room  \
0                       19.2                44.790000   
1                       19.2                44.722500   
2                       19.2                44.626667   
3                       19.2                44.590000   
4                       19.2                44.530000   

   Temperature_in_Laundryroom  Humidity_in_Laundary_room  \
0                       19.79                  44.730000   
1                       19.79                  44.790000   
2                       19.79                  44.933333   
3                       19.79                  45.000000   
4                       19.79                  45.000000   

   Temperature_in_office  Humidity_in_office  Temperature_in_Bathroom  ...  \
0              19.000000           45.566667                17.166667  ...   
1              19.000000           45.992500                17.166667  ...   
2              18.926667           45.890000                17.166667  ...   
3              18.890000           45.723333                17.166667  ...   
4              18.890000           45.530000                17.200000  ...   

   Temperature_in_Parent_room  Humidity_in_parent_room  Temperature_Outside  \
0                   17.033333                    45.53                 6.60   
1                   17.066667                    45.56                 6.48   
2                   17.000000                    45.50                 6.37   
3                   17.000000                    45.40                 6.25   
4                   17.000000                    45.40                 6.13   

   Pressure_mm_Hg  Humidiity_Outside  Windspeed  Visibility  Tdewpoint  \
0           733.5               92.0   7.000000   63.000000        5.3   
1           733.6               92.0   6.666667   59.166667        5.2   
2           733.7               92.0   6.333333   55.333333        5.1   
3           733.8               92.0   6.000000   51.500000        5.0   
4           733.9               92.0   5.666667   47.666667        4.9   

   Random_variable_1  Random_variable_2  
0          13.275433          13.275433  
1          18.606195          18.606195  
2          28.642668          28.642668  
3          45.410390          45.410390  
4          10.084097          10.084097  

[5 rows x 27 columns]

In [10]:
#Using min max scaler to normalize the dataset to a common scale
from sklearn.preprocessing import MinMaxScaler
scaler=MinMaxScaler()

#normalizing and droping some Features
norm=pd.DataFrame(scaler.fit_transform(new), columns=new.columns)
#dropping light
features=norm.drop(columns=['Appliances'])
Appliances=norm['Appliances']


In [11]:
norm.head()

Appliances  Temperature_in_Kitchen  Humidity_in_Kitchen  \
0    0.046729                 0.32735             0.566187   
1    0.046729                 0.32735             0.541326   
2    0.037383                 0.32735             0.530502   
3    0.037383                 0.32735             0.524080   
4    0.046729                 0.32735             0.531419   

   Temperature_in_Livingroom  Humidity_in_Living_room  \
0                   0.225345                 0.684038   
1                   0.225345                 0.682140   
2                   0.225345                 0.679445   
3                   0.225345                 0.678414   
4                   0.225345                 0.676727   

   Temperature_in_Laundryroom  Humidity_in_Laundary_room  \
0                    0.215188                   0.746066   
1                    0.215188                   0.748871   
2                    0.215188                   0.755569   
3                    0.215188                   0.758685   
4                    0.215188                   0.758685   

   Temperature_in_office  Humidity_in_office  Temperature_in_Bathroom  ...  \
0               0.351351            0.764262                 0.175506  ...   
1               0.351351            0.782437                 0.175506  ...   
2               0.344745            0.778062                 0.175506  ...   
3               0.341441            0.770949                 0.175506  ...   
4               0.341441            0.762697                 0.178691  ...   

   Temperature_in_Parent_room  Humidity_in_parent_room  Temperature_Outside  \
0                    0.223032                 0.677290             0.372990   
1                    0.226500                 0.678532             0.369132   
2                    0.219563                 0.676049             0.365595   
3                    0.219563                 0.671909             0.361736   
4                    0.219563                 0.671909             0.357878   

   Pressure_mm_Hg  Humidiity_Outside  Windspeed  Visibility  Tdewpoint  \
0        0.097674           0.894737   0.500000    0.953846   0.538462   
1        0.100000           0.894737   0.476190    0.894872   0.533937   
2        0.102326           0.894737   0.452381    0.835897   0.529412   
3        0.104651           0.894737   0.428571    0.776923   0.524887   
4        0.106977           0.894737   0.404762    0.717949   0.520362   

   Random_variable_1  Random_variable_2  
0           0.265449           0.265449  
1           0.372083           0.372083  
2           0.572848           0.572848  
3           0.908261           0.908261  
4           0.201611           0.201611  

[5 rows x 27 columns]

In [12]:
features.head()

Temperature_in_Kitchen  Humidity_in_Kitchen  Temperature_in_Livingroom  \
0                 0.32735             0.566187                   0.225345   
1                 0.32735             0.541326                   0.225345   
2                 0.32735             0.530502                   0.225345   
3                 0.32735             0.524080                   0.225345   
4                 0.32735             0.531419                   0.225345   

   Humidity_in_Living_room  Temperature_in_Laundryroom  \
0                 0.684038                    0.215188   
1                 0.682140                    0.215188   
2                 0.679445                    0.215188   
3                 0.678414                    0.215188   
4                 0.676727                    0.215188   

   Humidity_in_Laundary_room  Temperature_in_office  Humidity_in_office  \
0                   0.746066               0.351351            0.764262   
1                   0.748871               0.351351            0.782437   
2                   0.755569               0.344745            0.778062   
3                   0.758685               0.341441            0.770949   
4                   0.758685               0.341441            0.762697   

   Temperature_in_Bathroom  Humidity_in_Bathroom  ...  \
0                 0.175506              0.381691  ...   
1                 0.175506              0.381691  ...   
2                 0.175506              0.380037  ...   
3                 0.175506              0.380037  ...   
4                 0.178691              0.380037  ...   

   Temperature_in_Parent_room  Humidity_in_parent_room  Temperature_Outside  \
0                    0.223032                 0.677290             0.372990   
1                    0.226500                 0.678532             0.369132   
2                    0.219563                 0.676049             0.365595   
3                    0.219563                 0.671909             0.361736   
4                    0.219563                 0.671909             0.357878   

   Pressure_mm_Hg  Humidiity_Outside  Windspeed  Visibility  Tdewpoint  \
0        0.097674           0.894737   0.500000    0.953846   0.538462   
1        0.100000           0.894737   0.476190    0.894872   0.533937   
2        0.102326           0.894737   0.452381    0.835897   0.529412   
3        0.104651           0.894737   0.428571    0.776923   0.524887   
4        0.106977           0.894737   0.404762    0.717949   0.520362   

   Random_variable_1  Random_variable_2  
0           0.265449           0.265449  
1           0.372083           0.372083  
2           0.572848           0.572848  
3           0.908261           0.908261  
4           0.201611           0.201611  

[5 rows x 26 columns]

In [13]:
#Splitting dataset into training and test dataset
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test=train_test_split(features,Appliances,test_size=0.3, random_state=42)

In [14]:
#Linear Regresssion
from sklearn.linear_model import LinearRegression
linear_model=LinearRegression()
#fit model to a training dataset
linear_model.fit(x_train, y_train)
#obtain prediction
predicted_values = linear_model.predict(x_test)

In [15]:
#For Mean Average Error(MAE)
from sklearn.metrics import mean_absolute_error
mae=mean_absolute_error(y_test,predicted_values)

#Residual sum of squares
rss=np.sum(np.square(y_test - predicted_values))

#Root mean square error
from sklearn.metrics import mean_squared_error
rmse=np.sqrt(mean_squared_error(y_test, predicted_values))
round(mae,3)

#R-squared
from sklearn.metrics import r2_score
r_squared=r2_score(y_test,predicted_values)

print('rmse','is',round(rmse,3))
print('r_squared', 'is', round(r_squared,3))
print('mae','is',round(mae,3))
print('rss','is',round(rss,3))

rmse is 0.088
r_squared is 0.149
mae is 0.05
rss is 45.345


In [16]:
#ridge regression
from sklearn.linear_model import Ridge
ridge_reg=Ridge(alpha=0.5)
ridge_reg.fit(x_train, y_train)
pred2 = ridge_reg.predict(x_test) 

#evaluating metrics
from sklearn.metrics import mean_absolute_error
mae=mean_absolute_error(y_test,pred2)

#Residual sum of squares
rss=np.sum(np.square(y_test - pred2))

#Root mean square error
from sklearn.metrics import mean_squared_error
rmse=np.sqrt(mean_squared_error(y_test, pred2))
round(mae,3)

#R-squared
from sklearn.metrics import r2_score
r_squared=r2_score(y_test,pred2)

print('rmse','is',round(rmse,3))
print('r_squared', 'is', round(r_squared,3))
print('mae','is',round(mae,3))
print('rss','is',round(rss,3))

rmse is 0.088
r_squared is 0.148
mae is 0.05
rss is 45.376


In [17]:
#lasso regression
from sklearn.linear_model import Lasso
lasso_reg=Lasso(alpha=0.001)
lasso_reg.fit(x_train, y_train)
pred3 = lasso_reg.predict(x_test) 

#evaluating metrics
from sklearn.metrics import mean_absolute_error
mae=mean_absolute_error(y_test,pred3)

#Residual sum of squares
rss=np.sum(np.square(y_test - pred3))

#Root mean square error
from sklearn.metrics import mean_squared_error
rmse=np.sqrt(mean_squared_error(y_test, pred3))
round(mae,3)

#R-squared
from sklearn.metrics import r2_score
r_squared=r2_score(y_test,pred3)

print('rmse','is',round(rmse,3))
print('r_squared', 'is', round(r_squared,3))
print('mae','is',round(mae,3))
print('rss','is',round(rss,3))

rmse is 0.094
r_squared is 0.027
mae is 0.055
rss is 51.853


__Test Questions__

__Q12__

In [18]:
#temperature in living room
x=df.iloc[:,5].values
x=x.reshape(-1,1)
x

array([[19.2       ],
       [19.2       ],
       [19.2       ],
       ...,
       [25.62857143],
       [25.414     ],
       [25.26428571]])

In [19]:
#temperature outside the building
y=df.iloc[:,13].values
y=y.reshape(-1,1)
y

array([[ 7.02666667],
       [ 6.83333333],
       [ 6.56      ],
       ...,
       [23.62666667],
       [22.43333333],
       [21.02666667]])

In [20]:
#Splitting dataset into training and test dataset
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test=train_test_split(x,y,test_size=0.3, random_state=42)

#Linear Regresssion
from sklearn.linear_model import LinearRegression
linear_model=LinearRegression()
x_train=x.reshape(-1,1)
y_train=y=y.reshape(-1,1)
#fit model to a training dataset
linear_model.fit(x_train, y_train)
#obtain prediction
predicted_values = linear_model.predict(x_test)

__Q12-Q16.__

In [21]:
#For Mean Average Error(MAE)
from sklearn.metrics import mean_absolute_error
mae=mean_absolute_error(y_test,predicted_values)

#Residual sum of squares
rss=np.sum(np.square(y_test - predicted_values))

#Root mean square error
from sklearn.metrics import mean_squared_error
rmse=np.sqrt(mean_squared_error(y_test, predicted_values))

#R-squared
from sklearn.metrics import r2_score
r_squared=r2_score(y_test,predicted_values)

print('Coefficient of Determination or R_squared', 'is', round(r_squared,2))
print(' Mean Absolute Error','is',round(mae,2))
print('Residual sum of squares','is',round(rss,2))
print('Root Mean Squared Error','is',round(rmse,2))

Coefficient of Determination or R_squared is 0.64
 Mean Absolute Error is 2.82
Residual sum of squares is 78029.81
Root Mean Squared Error is 3.63


__Coefficient of Determination or R_squared is 0.64
 Mean Absolute Error is 2.82
Residual sum of squares is 78029.81
Root Mean Squared Error is 3.63.__

__Q18__

In [22]:
#ridge regression
from sklearn.linear_model import Ridge
ridge_reg=Ridge(alpha=0.4)
ridge_reg.fit(x_train, y_train)
pred2 = ridge_reg.predict(x_test) 

#Root mean square error
from sklearn.metrics import mean_squared_error
rmse=np.sqrt(mean_squared_error(y_test, pred2))
round(mae,3)

print('rmse','is',round(rmse,3))

rmse is 3.63


__There is no change on the rmse value.__

__Q20.__

In [23]:
#lasso regression
from sklearn.linear_model import Lasso
lasso_reg=Lasso(alpha=0.001)
lasso_reg.fit(x_train, y_train)
pred3 = lasso_reg.predict(x_test) 

#Root mean square error
from sklearn.metrics import mean_squared_error
rmse=np.sqrt(mean_squared_error(y_test, pred3))

print('Root mean square error','is',round(rmse,3))

Root mean square error is 3.63


__RMSE=3.63.__

## END OF NOTEBOOK